pipeline/pipeline.go

```diff
@@ -487,24 +530,27 @@ func (p *Pipeline) maintenance() {
 		p.antispamer.maintenance()
 		p.metricsHolder.maintenance()
 
-		totalCommitted := p.totalCommitted.Load()
-		deltaCommitted := int(totalCommitted - lastCommitted)
-
-		totalSize := p.totalSize.Load()
-		deltaSize := int(totalSize - lastSize)
+		// set new values and get the diffs
+		deltaInputEvents := inputEvents.updateValue(p.inputEvents.Load())
+		deltaInputSize := inputSize.updateValue(p.inputSize.Load())
+		deltaOutputEvents := outputEvents.updateValue(p.outputEvents.Load())
+		deltaOutputSize := outputSize.updateValue(p.outputSize.Load())
 
-		rate := int(float64(deltaCommitted) * float64(time.Second) / float64(interval))
-		rateMb := float64(deltaSize) * float64(time.Second) / float64(interval) / 1024 / 1024
-
-		tc := totalCommitted
-		if totalCommitted == 0 {
-			tc = 1
-		}
+		stats.GetCounter(p.subsystemName(), inputEventsCountMetric).Add(deltaInputEvents)
+		stats.GetCounter(p.subsystemName(), inputEventsSizeMetric).Add(deltaInputSize)
+		stats.GetCounter(p.subsystemName(), outputEventsCountMetric).Add(deltaOutputEvents)
+		stats.GetCounter(p.subsystemName(), outputEventsSizeMetric).Add(deltaOutputSize)
 
-		p.logger.Infof("%q pipeline stats interval=%ds, active procs=%d/%d, queue=%d/%d, out=%d|%.1fMb, rate=%d/s|%.1fMb/s, total=%d|%.1fMb, avg size=%d, max size=%d", p.Name, interval/time.Second, p.activeProcs.Load(), p.procCount.Load(), p.settings.Capacity-p.eventPool.freeEventsCount, p.settings.Capacity, deltaCommitted, float64(deltaSize)/1024.0/1024.0, rate, rateMb, totalCommitted, float64(totalSize)/1024.0/1024.0, totalSize/tc, p.maxSize)
+		rate := int(deltaInputEvents * float64(time.Second) / float64(interval))
+		rateMb := deltaInputSize * float64(time.Second) / float64(interval) / 1024 / 1024
+		tc := int64(math.Max(float64(inputEvents.get()), 1))
 
-		lastCommitted = totalCommitted
-		lastSize = totalSize
+		p.logger.Infof(`%q pipeline stats interval=%ds, active procs=%d/%d, queue=%d/%d, out=%d|%.1fMb,`+
```

Comment:
maybe move this in function?

Reply:
done, please take a look before i resolve this